### PR TITLE
refactor: Add AMIS GitHub Actions Workflow

### PR DESCRIPTION
**Pull request by jwagantall**
Add initial AMIS GHA workflow.
These workflow is migrated from CircleCI.

Signed-off-by: Jessica Wagantall <jwagantall@linuxfoundation.org>

## Additional Information

- [ ] This change is backwards-breaking

### CODE REVIEW

File: .github/workflows/amis-workflow.yml [hunk 1]
```diff
@@ -0,0 +1,143 @@
...
+    #  - ami-hold
+    if: ${{ needs.path_filter.outputs.should_skip != 'true' }}
+    name: publish-amis-to-marketplace job
+    runs-on: ubuntu-latest
+    env:
+      MAGMA_ROOT: "${{ github.workspace }}"
+      CODE_DIR: "${{ github.workspace }}/experimental/cloudstrapper"
+      VARS_DIR: "${{ github.workspace }}/experimental/cloudstrapper/playbooks/roles/vars"
+      WORK_DIR: "${{ github.workspace }}/experimental/cloudstrapper/playbooks"
+      AWS_ACCOUNT: "LF"
```
Comment (quentinDERORY) [inline .github/workflows/amis-workflow.yml]: On cron, it should be FB and on v1.* branches it should be LF. 
And please default it to FB while testing

Comment (jwagantall) [reply .github/workflows/amis-workflow.yml]: thanks!.. working on in now..

**Comment by quentinDERORY** [comment]
The original sed command for adding the AWS credentials was :`sed -i -e "s@awsAccessKey:@& ${AWS_ACCESS_KEY_AMI_QUENTIN}@1"  $VARS_DIR/secrets.yaml`
If it does not work with github action syntax try using single quote instead of double quote.

**Comment by jwagantall** [comment]
Thanks @quentinDERORY , i think this was my mistake but is resolved now past the secrets.

File: .github/workflows/amis-workflow.yml [hunk 2]
```diff
@@ -0,0 +1,156 @@
...
+            sed -i -e '/^stackCloudstrapper: /s/:.*$/: publish-amis-to-marketplace-stack-cloustrapper/' ${{ env.VARS_DIR }}/defaults.yaml
+            sed -i -e '/^devOpsAmi: /s/:.*$/: cloudstrapper-'"$VERSION"'/' ${{ env.VARS_DIR }}/defaults.yaml
+            sed -i -e '/^buildUbuntuAmi: /s/:.*$/: ami-09e67e426f25ce0d7/' ${{ env.VARS_DIR }}/build.yaml
+            sed -i -e '/^buildAgwVersion: /s/:.*$/: '"$GIT_REF"'/' ${{ env.VARS_DIR }}/build.yaml
+            sed -i -e '/^buildAgwPackage: /s/:.*$/: '"$PACKAGE_VERSION"'/' ${{ env.VARS_DIR }}/build.yaml
+            sed -i -e '/^taggedVersion: /s/:.*$/: '"$VERSION"'/' ${{ env.VARS_DIR }}/build.yaml
+      - name: Generate Cloudstrapper AMI
+        timeout-minutes: 30
+        run: |
+            echo "DEFAULTS"
```
Comment (jwagantall) [inline .github/workflows/amis-workflow.yml:89]: @quentinDERORY  my next error could also be related to the files created in the previous steps. 
Can you please help me confirming that the output of the vars, build and cluster yaml files is what we expect?

https://github.com/magma/magma/runs/3900359282?check_suite_focus=true

Comment (jwagantall) [reply .github/workflows/amis-workflow.yml:89]: oh..one sec.. i think i moved forward.. let me try one more update..

**Comment by jwagantall** [comment]
@quentinDERORY , do you know if this error makes sense or is it due to bad data passed through?
https://github.com/magma/magma/runs/3909661789?check_suite_focus=true

File: .github/workflows/amis-workflow.yml [hunk 3]
```diff
@@ -0,0 +1,157 @@
...
+          SLACK_ICON_EMOJI: ":heavy_check_mark:"
+          SLACK_COLOR: "#00FF00"
+          SLACK_FOOTER: ' '
+      - name: Propagate variables for AGW AMI build
+        run: |
+            sed -i -e '/^buildAgwAmiName: /s/:.*$/: agw-ami-'"$VERSION"'/' ${{ env.VARS_DIR }}/build.yaml
+            sed -i -e '/^buildGwTagName: /s/:.*$/: publishAmisToMarketplaceAgw/' ${{ env.VARS_DIR }}/build.yaml
+            sed -i -e '/^packageRepoHost: /s/:.*$/: '"${{ env.PACKAGE_REPO_HOST }}"'/' ${{ env.VARS_DIR }}/build.yaml
+            sed -i -e '/^awsAgwAmi: /s/:.*$/: ami-09e67e426f25ce0d7/' ${{ env.VARS_DIR }}/cluster.yaml
+            sed -i -e '/^buildAgwVersion: /s/:.*$/: '"refs\/heads\/${{ github.base_ref }}"'/' ${{ env.VARS_DIR }}/build.yaml
```
Comment (quentinDERORY) [inline .github/workflows/amis-workflow.yml:119]: Can you add a todo comment that is a workarround :)

File: .github/workflows/amis-workflow.yml [hunk 4]
```diff
@@ -0,0 +1,157 @@
...
+  schedule:
+    - cron: "0 0 * * 0"
+  # Add pull_request for testing
+  pull_request:
+    branches:
+      - master
+jobs:
+  publish-amis-to-marketplace:
+    # needs:
+    #  - release-hold
```
Comment (quentinDERORY) [inline .github/workflows/amis-workflow.yml]: remove commented code here

File: .github/workflows/amis-workflow.yml [hunk 5]
```diff
@@ -0,0 +1,157 @@
+---
+name: amis-workflow
+
+on:  # yamllint disable-line rule:truthy
+  push:
+    branches:
+      - 'v*'
```
Comment (quentinDERORY) [inline .github/workflows/amis-workflow.yml]: Can your replace with workflow dispatch as when we push to v1.*  packages are not published to production repo right away.

File: .github/workflows/amis-workflow.yml [hunk 6]
```diff
@@ -0,0 +1,157 @@
...
+      - uses: actions/setup-python@v2
+        with:
+          python-version: '3.7'
+      - name: Install Dependencies
+        run: |
+            pip install ansible awscli boto3
+            sudo apt-get update
+      - name: Determine AWS credentials to use
+        run: |
+            if ${{ github.ref == 'refs/heads/v1.*' && github.event_name != 'schedule' }}; then
```
Comment (quentinDERORY) [inline .github/workflows/amis-workflow.yml]: Need to change logic to work with workflow dispatch

File: .github/workflows/amis-workflow.yml [hunk 7]
```diff
@@ -0,0 +1,157 @@
...
+        run: |
+            ansible-playbook ${{ env.WORK_DIR }}/agw-provision.yaml -e "idSite=DevOps" -e "idGw=publishAmisToMarketplaceAgw" -e "dirLocalInventory=${{ env.VARS_DIR }}" --tags infra,inventory  -e "agwDevops=1" --skip-tags createBridge,cleanupBridge,cleanupNet
+            echo "Waiting one minute for the instance to boot up."
+            sleep 60
+            ansible-playbook ${{ env.WORK_DIR }}/ami-configure.yaml -i "${{ env.VARS_DIR }}/common_instance_aws_ec2.yaml" -e "dirLocalInventory=${{ env.VARS_DIR }}" -e "aminode=tag_Name_publishAmisToMarketplaceAgw" -e "ansible_python_interpreter=/usr/bin/python3" -u ubuntu
+            ansible-playbook ${{ env.WORK_DIR }}/ami-init.yaml -e "dirLocalInventory=${{ env.VARS_DIR }}"
+      - name: Export qcow2 image for fb aws account in s3
+        timeout-minutes: 30
+        run: |
+            if [ "<<parameters.aws_account>>" = "FB" ]; then
```
Comment (quentinDERORY) [inline .github/workflows/amis-workflow.yml]: this is circle ci syntax here

File: .github/workflows/amis-workflow.yml [hunk 8]
```diff
@@ -0,0 +1,157 @@
...
+              echo AWS_ACCOUNT="FB" >> $GITHUB_ENV
+            fi
+            # Default to FB for testing, delete this line before merging
+            echo AWS_ACCOUNT="FB" >> $GITHUB_ENV
+      - name: Propagate AWS credentials to ansible and create version
+        run: |
+            if [ "${{ env.AWS_ACCOUNT }}" = "FB" ]; then
+              sed -i -e "s@awsAccessKey:@& ${{ secrets.FB_AWS_ACCESS_KEY }}@1"  ${{ env.VARS_DIR }}/secrets.yaml
+              sed -i -e "s@awsSecretKey:@& ${{ secrets.FB_AWS_SECRET_ACCESS_KEY }}@1"  ${{ env.VARS_DIR }}/secrets.yaml
+              echo VERSION="1.6.0-${{ github.sha }}" >> $GITHUB_ENV
```
Comment (quentinDERORY) [inline .github/workflows/amis-workflow.yml]: can you bump this 1.7

File: .github/workflows/amis-workflow.yml [hunk 9]
```diff
@@ -0,0 +1,157 @@
...
+              echo PACKAGE_VERSION="1.6.0" >> $GITHUB_ENV
```
Comment (quentinDERORY) [inline .github/workflows/amis-workflow.yml]: here too

File: .github/workflows/amis-workflow.yml [hunk 10]
```diff
@@ -0,0 +1,157 @@
...
+              # When focal-ci is clean we will be able to dynamically fetch latest ci debian packages
+            elif  [ "${{ env.AWS_ACCOUNT }}" = "LF" ]; then
+              sed -i -e "s@awsAccessKey:@& ${{ secrets.LF_AWS_ACCESS_KEY }}@1"  ${{ env.VARS_DIR }}/secrets.yaml
+              sed -i -e "s@awsSecretKey:@& ${{ secrets.LF_AWS_SECRET_ACCESS_KEY }}@1"  ${{ env.VARS_DIR }}/secrets.yaml
+              echo PACKAGE_VERSION="1.7.0" >> $GITHUB_ENV
```
Comment (quentinDERORY) [inline .github/workflows/amis-workflow.yml]: when triggered with workflow dispatch on a branch, could you parse the branch name v1.x.x to get 1.x.x here

**Review by quentinDERORY — CHANGES_REQUESTED** [review]
See comments

File: .github/workflows/amis-workflow.yml (hunk 4 above)
Comment (jwagantall) [reply .github/workflows/amis-workflow.yml]: done

File: .github/workflows/amis-workflow.yml (hunk 8 above)
Comment (jwagantall) [reply .github/workflows/amis-workflow.yml]: sure! done

File: .github/workflows/amis-workflow.yml (hunk 9 above)
Comment (jwagantall) [reply .github/workflows/amis-workflow.yml]: done

File: .github/workflows/amis-workflow.yml (hunk 3 above)
Comment (jwagantall) [reply .github/workflows/amis-workflow.yml:119]: oh yes!

File: .github/workflows/amis-workflow.yml (hunk 7 above)
Comment (jwagantall) [reply .github/workflows/amis-workflow.yml]: oops.. fixed

File: .github/workflows/amis-workflow.yml (hunk 5 above)
Comment (jwagantall) [reply .github/workflows/amis-workflow.yml]: thanks!.. sorry I am quite new to how the workflow dispatch works (I think I have only used it with inputs)
So.. will something like this be acceptable?

  workflow_dispatch:
    branches:
      - 'v1.*'

File: .github/workflows/amis-workflow.yml (hunk 6 above)
Comment (jwagantall) [reply .github/workflows/amis-workflow.yml]: done

File: .github/workflows/amis-workflow.yml (hunk 10 above)
Comment (jwagantall) [reply .github/workflows/amis-workflow.yml]: I wrote a potential way to do this, but will need to test

File: .github/workflows/amis-workflow.yml [hunk 11]
```diff
@@ -0,0 +1,156 @@
...
+name: amis-workflow
+
+on:  # yamllint disable-line rule:truthy
+  workflow_dispatch:
+    branches:
+      - 'v1.*'
+  schedule:
+    - cron: "0 0 * * 0"
+  # Add pull_request for testing
+  pull_request:
```
Comment (jwagantall) [inline .github/workflows/amis-workflow.yml]: @quentinDERORY  do we need to remove the pull_request event?

The AWS account to use step (line 39) only supports Schedule or Worflow_dispatch

File: .github/workflows/amis-workflow.yml (hunk 5 above)
Comment (quentinDERORY) [reply .github/workflows/amis-workflow.yml]: just specify `workflow_dispatch: `, the ui will display a branch selector in github
you cannot specify filter on the branches, you would have to do it as a step if you want.

File: .github/workflows/amis-workflow.yml (hunk 11 above)
Comment (quentinDERORY) [reply .github/workflows/amis-workflow.yml]: yes remove pull_request event 👍

Comment (jwagantall) [reply .github/workflows/amis-workflow.yml]: Thanks! I will also remove the lines that I had to force AWS_ACCOUNT = "FB"
that was used for testing too..

**Comment by jwagantall** [comment]
@quentinDERORY  I am able to see the workflow run option in my fork now 
https://github.com/jwagantall/magma/actions/runs/1473458471 (It will fail since I don't have magma's secrets)

**Review by quentinDERORY — APPROVED** [review]

**Pull request merged**